tpls/desul/include/desul/atomics/Lock_Array.hpp

```diff
@@ -17,6 +17,11 @@ SPDX-License-Identifier: (BSD-3-Clause)
 #ifdef DESUL_HAVE_HIP_ATOMICS
 #include <desul/atomics/Lock_Array_HIP.hpp>
 #endif
+// Different from DESUL_HAVE_SYCL_ATOMICS because we need the declaration of
+// the sycl::device_global variables in both compile passes.
+#ifdef SYCL_LANGUAGE_VERSION
```

Comment:
I was looking into the `<desul/atomicsMacros.hpp>`  and it just occurred to me:
https://github.com/kokkos/kokkos/blob/b0cc0f63507d28d7c393bf18c33bd07341e319cf/tpls/desul/include/desul/atomics/Macros.hpp#L42-L44
Why are we doing `#ifdef __SYCL_DEVICE_ONLY__` and not `#ifdef SYCL_LANGUAGE_VERSION` there?
For CUDA and HIP we are guarding for `__CUDACC__` and `__HIPCC__`, not `__CUDA_ARCH__` and `__HIP_DEVICE_COMPILE__`.  Would that resolve the issue you are seeing?